xdem/coreg.py

```diff
@@ -1712,3 +1714,144 @@ def _to_matrix_func(self) -> np.ndarray:
+
+    bounds, resolution = _transform_to_bounds_and_res(dem.shape, transform)
+
+    # If a centroid was not given, default to the bottom left corner.
+    if centroid is None:
+        centroid = (bounds.left, bounds.bottom, 0.0)
+    else:
+        assert len(centroid) == 3, f"Expected centroid to be 3D X/Y/Z coordinate. Got shape of {len(centroid)}"
+
+    # Shift the coordinates to centre around the centroid.
+    x_coords -= centroid[0]
+    y_coords -= centroid[1]
+
+    # Create a point cloud of X/Y/Z coordinates
+    point_cloud = np.dstack((x_coords, y_coords, filled_dem))
+
+    # Shift the Z components by the centroid.
+    point_cloud[:, 2] -= centroid[2]
+
+    if invert:
+        matrix = invert_matrix(matrix)
+
+    # Transform the point cloud using the matrix.
+    transformed_points = cv2.perspectiveTransform(
+        point_cloud.reshape((1, -1, 3)),
+        matrix,
+    ).reshape(point_cloud.shape)
+
+    # Estimate the vertical difference of old and new point cloud elevations.
+    deramp = deramping(
+        (point_cloud[:, :, 2] - transformed_points[:, :, 2])[~nan_mask].flatten(),
+        point_cloud[:, :, 0][~nan_mask].flatten(),
+        point_cloud[:, :, 1][~nan_mask].flatten(),
+        degree=1
+    )
+    # Shift the elevation values of the soon-to-be-warped DEM.
+    filled_dem -= deramp(x_coords, y_coords)
+
+    # Create gap-free arrays of x and y coordinates to be converted into index coordinates.
+    x_inds = rio.fill.fillnodata(transformed_points[:, :, 0].copy(), mask=(~nan_mask).astype("uint8"))
+    y_inds = rio.fill.fillnodata(transformed_points[:, :, 1].copy(), mask=(~nan_mask).astype("uint8"))
+
+    # Divide the coordinates by the resolution to create index coordinates.
+    x_inds /= resolution
+    y_inds /= resolution
+    # Shift the x coords so that bounds.left is equivalent to xindex -0.5
+    x_inds -= x_coords.min() / resolution
+    # Shift the y coords so that bounds.top is equivalent to yindex -0.5
+    y_inds = (y_coords.max() / resolution) - y_inds
+
+    # Create a skimage-compatible array of the new index coordinates that the pixels shall have after warping.
+    inds = np.vstack((y_inds.reshape((1,) + y_inds.shape), x_inds.reshape((1,) + x_inds.shape)))
+
+    # Warp the DEM
+    transformed_dem = skimage.transform.warp(
+        filled_dem,
+        inds,
+        order=resampling_order,
+        mode="constant",
+        cval=0,
+        preserve_range=True
+    )
+    # Warp the NaN mask, setting true to all values outside the new frame.
+    tr_nan_mask = skimage.transform.warp(
+        nan_mask.astype("uint8"),
+        inds,
+        order=resampling_order,
+        mode="constant",
+        cval=1,
+        preserve_range=True
+    ) > 0.5  # Due to different interpolation approaches, everything above 0.5 is assumed to be 1 (True)
+
+    # Apply the transformed nan_mask
+    transformed_dem[tr_nan_mask] = np.nan
```

Comment:
Sorry, but could you elaborate here? The warp function has a fill value of 1 (`cval=1`), meaning all values outside the frame (near the border) should be 1 (True).

Reply:
What I'm trying to say is that when you create a `filled_dem` with `-9999` at the NaN values (or anything else) in order to run `skimage.transform.warp`, those will be used in the interpolations with resampling_order> 0. Because of this, it will create a bias for non-NaN values that are close to NaN values (within 3 pixels for a resampling order of 3, for instance).
As a consequence, it could be a good idea to extend the NaN mask using a binary dilation with the same order than the resampling method used (like it is done when deriving slope, for instance).
Or maybe I am missing something? (I didn't use skimage much)

Reply:
But if the nan mask is dilated, won't it be smaller and smaller for every iteration? So: 
```python
while True:
    dem = apply_matrix(dem, transform, empty_matrix)
```
Would eventually mean that `dem` is empty, right?

Reply:
Yes, that's the behaviour! GDAL behaves similarly. If you resample a raster with resampling_order>0 interpolation 10 times, the NaNs will propagate into bigger and bigger data gaps and eventually you will be left with nothing.
It should be only an option for sure, but someone might want to avoid using data at the edges after resampling?

I'm still not sure I get 100% how skimage deals with the nodata here: are the filled values of "-9999" actually used in the resampling and then masked later? This would create strong edge effects by including the -9999 in the resampling.

Reply:
Yes, definitely something worth checking and testing. Have you tried running the function on a DEM with gaps to see how it behaves?

Reply:
I've added a `dilate_mask` argument in 7095024, @rhugonnet.

I also changed the temporary nodata value to be the nanmedian (instead of -9999), which I guess should reduce the risk of extreme outliers.

The tests include nans, @adehecq, so that does not seem to be a problem.